go-controller/pkg/util/kube.go

```diff
@@ -719,55 +733,134 @@ func NoHostSubnet(node *v1.Node) bool {
-			for _, ip := range endpoint.Addresses {
-				endpointsAddress.Insert(utilnet.ParseIPSloppy(ip).String())
-			}
-		})
+		endpoints = append(endpoints, endpointSlice.Endpoints...)
 	}
+	return endpoints
+}
+
+// GetEligibleEndpointAddressesWithPreCondition first applies the (optionally) provided condition function condFn
+// to filter endpoints from the given list of endpoint slices; from the resulting list,
+// it returns the IP addresses of eligible endpoints.
+func GetEligibleEndpointAddressesWithPreCondition(endpointSlices []*discovery.EndpointSlice, service *kapi.Service, condFn func(discovery.Endpoint) bool) sets.Set[string] {
+	endpointsAddress := sets.New[string]()
+
+	endpoints := getAllEndpointsFromEndpointSlices(endpointSlices...)
+
+	forEachSelectedAndEligibleEndpoint(endpoints, service, condFn, func(endpoint discovery.Endpoint) {
+		for _, ip := range endpoint.Addresses {
+			endpointsAddress.Insert(utilnet.ParseIPSloppy(ip).String())
+		}
+	})
+
 	return endpointsAddress
 }
 
-// GetEndpointAddresses returns a list of IP addresses of all eligible endpoints in the given endpoint slice.
-func GetEndpointAddresses(endpointSlices []*discovery.EndpointSlice, service *kapi.Service) sets.Set[string] {
-	return GetEndpointAddressesWithCondition(endpointSlices, service, nil)
+// GetEligibleEndpointAddresses returns a list of IP addresses of all eligible endpoints in the given endpoint slice.
+func GetEligibleEndpointAddresses(endpointSlices []*discovery.EndpointSlice, service *kapi.Service) sets.Set[string] {
+	return GetEligibleEndpointAddressesWithPreCondition(endpointSlices, service, nil)
 }
 
-// GetLocalEndpointAddresses returns a list of endpoints that are local to the specified node
-func GetLocalEndpointAddresses(endpointSlices []*discovery.EndpointSlice, service *kapi.Service, nodeName string) sets.Set[string] {
-	return GetEndpointAddressesWithCondition(endpointSlices, service, func(endpoint discovery.Endpoint) bool {
+// GetLocalEligibleEndpointAddresses returns a list of endpoints that are local to the specified node and are eligible.
+func GetLocalEligibleEndpointAddresses(endpointSlices []*discovery.EndpointSlice, service *kapi.Service, nodeName string) sets.Set[string] {
+	return GetEligibleEndpointAddressesWithPreCondition(endpointSlices, service, func(endpoint discovery.Endpoint) bool {
 		return endpoint.NodeName != nil && *endpoint.NodeName == nodeName
 	})
 }
 
+// DoesEndpointSliceContainEndpoint returns true if the endpointslice
+// contains an endpoint with the given IP/Port/Protocol and this endpoint is considered eligible
+func DoesEndpointSliceContainEligibleEndpoint(endpointSlice *discovery.EndpointSlice,
+	epIP string, epPort int32, protocol kapi.Protocol, service *kapi.Service) bool {
+	var containsEligibleEndpoint bool
+	endpoints := getAllEndpointsFromEndpointSlices(endpointSlice)
+	forEachSelectedAndEligibleEndpoint(endpoints, service, nil, func(ep discovery.Endpoint) {
+		if !containsEligibleEndpoint {
```

Comment:
Wouldn't it be better to write this as a condition function?

Reply:
I wouldn't do it for the same reason I was explaining to Jacob in https://github.com/ovn-org/ovn-kubernetes/pull/4072#discussion_r1443073083
We want to first do the endpoint selection based on readiness and then check whether the given endpoint is in there. I would limit `condFn` to a filter against local/non-local endpoints. 
If I include the check for the endpoint inside the logic for endpoint selection, I might not always get the correct result.

For instance, let's say we have an endpointslice with two endpoints:
- `ep1` on node1: ready=false, serving=true, terminating=true
- `ep2` on node2: ready=true, serving=true, terminating=false

I run `DoesEndpointSliceContainEligibleEndpoint` on the endpointslice above and against `ep1`
Let's say the service is of type cluster IP, so ovnkube-controller (`GetLbEndpoints`) picked for this service only the `ep2` endpoint, since it's the only one that is ready. The correct output for `DoesEndpointSliceContainEligibleEndpoint` is `false`.

In the current implementation of `DoesEndpointSliceContainEligibleEndpoint`, first we do the eligible endpoint selection, which returns `ep2`, and then the provided `fn` will check whether `ep1` is in there. It'll return false.

If I move the code from `fn` to `condFn`, we first select only `ep1` in `condFn` and then we mark it as eligible since it's terminating and serving and we have no other endpoint to consider. It'll return true.

Reply:
Understood.

I think that merits a comment with a hint, the explanation doesn't need to as extensive as this one.

I am just concerned that we might not be being explicit eoough with this problematic and we might have bugs int he future.

Would it make sense to instead of `forEachSelectedAndEligibleEndpoint` we just have
```
func getSelectedSelectedEligibleEndpoint(endpoints []discovery.Endpoint, service *kapi.Service, condFn func(ep discovery.Endpoint) bool) []discovery.Endpoint
```
and then any user of this function would have to iterate again for any post filtrering/processing? Because that last iteration would be unavoidable anyway, it may well be done by the user of the function instead of within the function itself. And the function itself would not introduce any confusion between pre/post filtering, it would be just pre-filtering.

Reply:
Sorry if I am going in circles here.